workflow.php

```diff
@@ -87,136 +61,66 @@ public function getUserRecord($userId){
-				break;
+		else {
+			$url = 'https://pitt-illiad-oclc-org.pitt.idm.oclc.org/illiad/illiad.dll/OpenURL?';
 		}
-		foreach ($map as $k => $v) {
-			$url .= '&'.$k.'='.$params[$v];
+		$userQueryString = '';
+		//sanitize user-generated query string parameters
+		foreach($_GET as $k=>$v) {
+			//php replaces query string parameter keys containing dots as underscores, so rft.title becomes rft_title. Change that back with str_replace, but only for rft keys. rfr, for example, uses an underscore natively, like rfr_id
+			$userQueryString .= str_replace("rft_", "rft.", urlencode($k)).'='.urlencode($v).'&';
 		}
+		//the last param should not be followed by &
+		$url .= rtrim($userQueryString,'&');
 		return $url;
 	}
 }
 
+//These Alma user groups aren't permitted to place external requests
+$noExternalBorrowing = array('LAWSPBORROWER','PATPURGE','PITTLIBASSIGNMENT','PROBLEM','ULSSPBORROWER','ULSSPRECIPROCAL','UPPROGRAM');
+
 //Get the patron's Pitt username
 $user = new Alma();
 $userId = $user->getUserId();
+$requestStatus = '';
 
 //if their account has all the info we need
-if ($user->getUserRecord($userId) && $user->getUserRecord($userId)->campus_code && $user->getUserRecord($userId)->user_group->value) {
-	//this user group isn't permitted to place external requests
-	if ($user->getUserRecord($userId)->user_group->value == 'UPPROGRAM') {
-		$requestStatus='program';
+if (($userRecord = $user->getUserRecord($userId)) && $userRecord->campus_code && $userRecord->user_group->value) {
+	//Find user's Alma usergroup and campus
+	//If the blank Campus dropdown menu option in Alma is saved for a user, their user campus_code object doesn't get a value, just a blank description property
+	if (!property_exists($userRecord->campus_code,'value')) {
+		$campus = 'BLANK';
 	}
-	else {
-		//continue on with the existing workflow
-		//Determine which Alma campus the user belongs to
-		$campus = $user->getUserRecord($userId)->campus_code->value;
-		//Does user have an ILLiad account? We'll check based on their campus and corresponding library system
-		$illiad = new Illiad($campus);
-		$illiadUserExists = $illiad->userExists($userId);
-		//construct a link to the appropriate ILLiad form
-		$illiadUrl = $illiad->buildUrl($campus, $userSubmittedParams);
-		//if they already have an ILLiad account
-		if ($illiadUserExists) {
-			//send them immediately to the ILLiad request form
-			//if not, the html instructions below will display by default
-			header("Location: $illiadUrl");
-		}
+	//Otherwise, just take the campus value straight from the user record 
+	else{
+		$campus = $userRecord->campus_code->value;
+	}
+	//Barco Law Library users do not participate in Illiad, so we'll handle them separately later 
+	if ($user->getUserRecord($userId)->campus_code->value==='LAW' && $requestStatus !== 'specialBorrower') {
```

Comment:
Why doesn't this also use `$userRecord`, and will it not throw warning if `->campus_code` does not have a `->value` property?

Reply:
Resolved by moving this inside the conditional that checks for the value property. No reason not to use $userRecord here. Switched. Thank you!